Automattic/sensei
# Prevent sending welcome email immediately on enrole in case their access starts in the future

Resolves https://github.com/Automattic/sensei-pro/issues/2572
(Partially)

## Proposed Changes
1. We have switched from hook `sensei_user_course_start` to `sensei_course_enrolment_status_changed` because `sensei_user_course_start ` hook is fired before the access start date is set. So it's not possible at that point to prevent or allow this email based on access. `sensei_course_enrolment_status_changed` is fired after everything is in place.

2. We have improved our base email generator class. Now it allows you to listen to multiple hooks. Previously you could only attach to one.

3. Updated the `sensei_send_emails` to contain the placeholders as well. This will allow us to add listener hook functions that take decisions based on relevant values.


## Testing Instructions


1. Disable your local sensei-pro
2. Create two Courses
3. Enroll as a student
4. Make sure your receive your welcome email immediately
5. Go to Sensei LMS -> Emails
6. Disable the welcome email in the student tab
7. Now enroll in another course
8. Make sure that you haven't received a welcome email this time
9. Enable Sensei Pro, then follow the instructions here https://github.com/Automattic/sensei-pro/pull/2586
10. Try it with WPML enabled and disabled.
11. Try it with a course that has more than one WPML translations. Make sure the email is sent only once. Not for each translation.

## New/Updated Hooks


* `sensei_send_emails` - Added `placeholders` param.

## Pre-Merge Checklist
